.teamcity/src/main/kotlin/model/FunctionalTestBucketGenerator.kt

```diff
@@ -82,42 +83,54 @@ data class FunctionalTestBucketWithSplitClasses(
 
 data class MultipleSubprojectsFunctionalTestBucket(
     val subprojects: List<String>,
-    val enableTD: Boolean
+    val parallelizationMethod: ParallelizationMethod?
 ) : FunctionalTestBucket {
     constructor(jsonObject: JSONObject) : this(
         jsonObject.getJSONArray("subprojects").map { it.toString() },
-        jsonObject.getBoolean("enableTD")
+        parseParallelization(jsonObject)
     )
 
     override fun toBuildTypeBucket(gradleSubprojectProvider: GradleSubprojectProvider): BuildTypeBucket {
         return SmallSubprojectBucket(
             subprojects.map { gradleSubprojectProvider.getSubprojectByName(it)!! },
-            enableTD
+            parallelizationMethod
         )
     }
+
+    companion object {
+        // Method is a companion function, so it can be used from the constructor
+        fun parseParallelization(jsonObject: JSONObject): ParallelizationMethod? {
+            return jsonObject.getJSONObject("parallelizationMethod")?.let { parallelizationMethod ->
+                parallelizationMethod.getString("name")?.let { parallelizationName ->
+                    when (parallelizationName) {
+                        ParallelizationMethod.TestDistribution.getName() -> ParallelizationMethod.TestDistribution()
+                        ParallelizationMethod.ParallelTesting.getName() -> ParallelizationMethod.ParallelTesting(parallelizationMethod.getIntValue("numberOfBuckets"))
+                        else -> throw IllegalArgumentException("Unknown parallelization method")
+                    }
+                }
+            }
+        }
+    }
 }
 
 fun BuildTypeBucket.toJsonBucket(): FunctionalTestBucket {
     return when (this) {
-        is SmallSubprojectBucket -> MultipleSubprojectsFunctionalTestBucket(subprojects.map { it.name }, enableTestDistribution)
+        is SmallSubprojectBucket -> MultipleSubprojectsFunctionalTestBucket(subprojects.map { it.name }, parallelizationMethod)
         is LargeSubprojectSplitBucket -> FunctionalTestBucketWithSplitClasses(subproject.name, number, classes.map { it.toPropertiesLine() }, include)
         else -> throw IllegalStateException("Unsupported type: ${this.javaClass}")
     }
 }
 
 class SubprojectTestClassTime(
     val subProject: GradleSubproject,
-    val testClassTimes: List<TestClassTime> = emptyList()
+    private val testClassTimes: List<TestClassTime> = emptyList()
 ) {
     val totalTime: Int = testClassTimes.sumOf { it.buildTimeMs }
 
-    fun split(expectedBucketNumber: Int, enableTestDistribution: Boolean = false): List<BuildTypeBucket> {
+    private fun splitIntoBuckets(expectedBucketNumber: Int): List<BuildTypeBucket> {
         return if (expectedBucketNumber == 1) {
             listOf(
-                SmallSubprojectBucket(
-                    listOf(subProject),
-                    enableTestDistribution
-                )
+                SmallSubprojectBucket(listOf(subProject), null)
```

Comment:
Looks like a mistake, as `null` signifies no parallelization (before there was the option to enable TD).
PTAL at the new function `split` if you wonder what's happening.